worlds/pokemon_emerald/__init__.py

```diff
@@ -0,0 +1,794 @@
+            convert_unrandomized_items_to_events("KeyItem")
+
+
+    def pre_fill(self):
+        # Items which are shuffled between their own locations
+        if self.multiworld.badges[self.player].value == RandomizeBadges.option_shuffle:
+            badge_locations = [location for location, _ in self.badge_shuffle_info]
+            badge_items = [item for _, item in self.badge_shuffle_info]
+
+            collection_state = self.multiworld.get_all_state(False)
+            if self.hm_shuffle_info is not None:
+                for _, item in self.hm_shuffle_info:
+                    collection_state.collect(item)
+                collection_state.sweep_for_events()
+
+            self.multiworld.random.shuffle(badge_locations)
+            self.multiworld.random.shuffle(badge_items)
+
+            fill_restrictive(self.multiworld, collection_state, badge_locations, badge_items, True, True)
+
+        if self.multiworld.hms[self.player].value == RandomizeHms.option_shuffle:
+            hm_locations = [location for location, _ in self.hm_shuffle_info]
+            hm_items = [item for _, item in self.hm_shuffle_info]
+
+            collection_state = self.multiworld.get_all_state(False)
+
+            self.multiworld.random.shuffle(hm_locations)
+            self.multiworld.random.shuffle(hm_items)
+
+            fill_restrictive(self.multiworld, collection_state, hm_locations, hm_items, True, True)
+
+
+    def generate_output(self, output_directory: str):
+        random = self.multiworld.per_slot_randoms[self.player]
+
+        def randomize_abilities():
+            # Creating list of potential abilities
+            ability_label_to_value = {ability.label.lower(): ability.ability_id for ability in emerald_data.abilities}
+
+            ability_blacklist_labels = {"cacophony"}
+            option_ability_blacklist = self.multiworld.ability_blacklist[self.player].value
+            if option_ability_blacklist is not None:
+                ability_blacklist_labels |= {ability_label.lower() for ability_label in option_ability_blacklist}
+
+            ability_blacklist = {ability_label_to_value[label] for label in ability_blacklist_labels}
+            ability_whitelist = [a.ability_id for a in emerald_data.abilities if a.ability_id not in ability_blacklist]
+
+            if self.multiworld.abilities[self.player].value == RandomizeAbilities.option_follow_evolutions:
+                already_modified: Set[int] = set()
+
+                # Loops through species and only tries to modify abilities if the pokemon has no pre-evolution
+                # or if the pre-evolution has already been modified. Then tries to modify all species that evolve
+                # from this one which have the same abilities.
+                # The outer while loop only runs three times for vanilla ordering: Once for a first pass, once for
+                # Hitmonlee/Hitmonchan, and once to verify that there's nothing left to do.
+                while True:
```

Comment:
It can't. The list of species is always in the same order, and their abilities are always vanilla until after this loop has run, so it will always modify species in the exact same order and exit after the third inner loop finds nothing to modify.

But to ease your mind further, I also tried

- Shuffling the order of the list on every iteration of the outer loop
- Reversing the order of the list
- Randomizing the evolutions of every species before the loops

In no case did the outer loop ever run more than 4 times. It theoretically could run more iterations for a randomized evo chain with a bunch of middle evolutions that are in exactly reverse order of their ids, but even then it will terminate.

I _suppose_ the `while len(evolutions) > 0` loop could run forever if a species were to evolve directly into itself, but that's not currently possible. But I guess I should protect it anyway.

Reply:
951ba3c makes it bulletproof even if evolution randomization is implemented poorly.